### PR TITLE
More accurate NPC hp overlay when max hp is known.

### PR DESCRIPTION
**Pull request by henke96**
I figured out how the server calculates the healthRatio it sends the client (which is healthRatio = 1 + (healthScale - 1) * health / maxHealth for health > 0 and 0 for health = 0).

I have tested this on quite a lot of different npcs, for some i've confirmed every single hp value. I've also tested on bosses with large hp bars and have not been able to disprove it at all.

Knowing how the server calculates the ratio means we can reverse the interval of possible hp that could've resulted in it. Which is exactly what i've done, i then show the middle between the max/min possible hps.

Note that the current hp display is incorrect even for NPCs with a 1:1 relation between possible hp and healthratio (those with max hp <= health scale).

### CODE REVIEW

**Comment by Gamer1120** [comment]
Line 192-194 could be a /*  ...   */ comment.

File: runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoOverlay.java [hunk 1]
```diff
@@ -143,21 +145,38 @@ public Dimension render(Graphics2D graphics)
...
 			if (lastMaxHealth != null)
 			{
+				int health = 0;
+				if (lastRatio > 0)
+				{
+					// This is the reverse of the calculation of healthRatio done by the server
+					// which is: healthRatio = 1 + (healthScale - 1) * health / maxHealth (if health > 0, 0 otherwise)
+					// It's able to recover the exact health if maxHealth <= healthScale.
+					int minHealth = (lastMaxHealth * (lastRatio - 1) - 1) / (lastHealthScale - 1) + 1;
+					int maxHealth = (lastMaxHealth * lastRatio - 1) / (lastHealthScale - 1);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoOverlay.java]: This looks like it would divide by zero

Comment (henke96) [reply runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoOverlay.java]: Hmm ye. That would be a 1 pixel wide health bar tho, but i can add a check for it.

**Comment by henke96** [comment]
Looked over the math again. Basically if healthScale is 1, the server calculation will give healthRatio = 1 for all healths except for 0. This means that the upper limit is basically infinity, which kinda explains the division by 0. I made a special case for this where it sets the upper limit to the NPCs max health.

Also added a missing special case for the lower limit.

**Comment by Nightfirecat** [comment]
If merged, I think this would fix #58? Needs rebase after deob purge.

**Comment by Nightfirecat** [comment]
@henke96 What's the status of this PR? Sounds like this PR still applies

**Pull request closed without merging**